presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,253 @@
+                        .getDynamicConjuncts().stream()
+                        .map(DynamicFilters.Descriptor::getId))
+                .collect(toImmutableList());
+        registerQuery(sqlQueryExecution.getQueryId(), sqlQueryExecution::getAllStages, dynamicFilterIds);
+    }
+
+    @VisibleForTesting
+    public synchronized void registerQuery(QueryId queryId, Supplier<List<StageInfo>> stageInfoSupplier, List<String> dynamicFilterIds)
+    {
+        queries.putIfAbsent(queryId, stageInfoSupplier);
+        dynamicFilterIds.forEach(filterId -> dynamicFilterSummaries.put(
+                SourceDescriptor.of(queryId, filterId),
+                Optional.empty()));
+    }
+
+    public synchronized void removeQuery(QueryId queryId)
+    {
+        dynamicFilterSummaries.keySet().removeIf(sourceDescriptor -> sourceDescriptor.getQueryId().equals(queryId));
+        queries.remove(queryId);
+    }
+
+    @VisibleForTesting
+    public synchronized void collectDynamicFilters()
+    {
+        List<QueryId> queriesWithDynamicFilters = dynamicFilterSummaries.entrySet()
+                .stream()
+                .filter(entry -> !entry.getValue().isPresent())
+                .map(entry -> entry.getKey().getQueryId())
+                .distinct()
+                .collect(toImmutableList());
+
+        for (QueryId queryId : queriesWithDynamicFilters) {
+            Supplier<List<StageInfo>> stageInfoSupplier = queries.get(queryId);
+            for (StageInfo stageInfo : stageInfoSupplier.get()) {
+                StageState stageState = stageInfo.getState();
+                // wait until stage has finished scheduling tasks
+                if (!stageState.noMoreTasks()) {
+                    continue;
+                }
+                List<TaskInfo> tasks = stageInfo.getTasks();
+                Map<String, List<Domain>> stageDynamicFilterDomains = tasks.stream()
+                        .map(taskInfo -> taskInfo.getTaskStatus().getDynamicFilterDomains())
+                        .flatMap(taskDynamicTupleDomains -> taskDynamicTupleDomains.entrySet().stream())
+                        .collect(groupingBy(Map.Entry::getKey, mapping(Map.Entry::getValue, toList())));
+
+                stageDynamicFilterDomains.entrySet().stream()
+                        // check if all tasks of a dynamic filter source have reported dynamic filter summary
+                        .filter(stageDomains -> stageDomains.getValue().size() == tasks.size())
+                        .forEach(stageDomains -> dynamicFilterSummaries.put(
+                                SourceDescriptor.of(queryId, stageDomains.getKey()),
+                                Optional.of(Domain.union(stageDomains.getValue()))));
+            }
+        }
+    }
+
+    @VisibleForTesting
+    public Optional<Domain> getSummary(QueryId queryId, String filterId)
+    {
+        Optional<Domain> domain = dynamicFilterSummaries.get(SourceDescriptor.of(queryId, filterId));
```

Comment:
use https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#getOrDefault-java.lang.Object-V- instead:
```
return dynamicFilterSummaries.getOrDefault(SourceDescriptor.of(queryId, filterId), Optional.empty());
```

Reply:
Removed Optional from map but kept it in this method definition to give cleaner API it's callers. Can't use `getOrDefault` when default value type is different from map value type.